src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -209,7 +211,14 @@ QueryPlanPtr MergeTreeDataSelectExecutor::readFromParts(
     const auto & primary_key = metadata_snapshot->getPrimaryKey();
     Names primary_key_columns = primary_key.column_names;
 
-    KeyCondition key_condition(query_info, context, primary_key_columns, primary_key.expression);
+    // query_info_for_index is a cloned SelectQueryInfo just for index
+    SelectQueryInfo query_info_for_index = query_info;
+    if (!metadata_snapshot->getColumns().getAliases().empty())
+    {
+        query_info_for_index.query = replaceAliasColumnsInFilter(query_info.query->clone(), metadata_snapshot->getColumns());
```

Comment:
By the way looks like this is done not for the first time, but it work put this info into the `SelectQueryInfo`?